### PR TITLE
[NEXMANAGE-314] Send response immediately for immediately scheduling

### PR DESCRIPTION
**Pull request by yengliong93**
_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
For all requests, the dispatcher agent should send back the acknowledge response ASAP. This PR updates the dispatcher to send back the acknowledgement response immediately before SOTA during the immediately scheduling.

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): **\<URL to be filled>**
- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [x] Changelogs are updated (or N/A if not customer visible)
- [ ] inbm/log_changes.txt are updated for potentially Validation-breaking log changes (or N/A if none)

### CODE MAINTAINABILITY
- [x] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [x] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [ ] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):
- [ ] Run 'go fmt' or format-python.sh as applicable.
- [ ] New/modified methods and functions should have type annotations on signatures as applicable
- [ ] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**
- [ ] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

### CODE REVIEW

**Comment by pillaiabhishek** [comment]
@yengliong93 @gblewis1, I have couple of questions:
1. I see that the changes for this change is only done for INBS cloud adapter. Could this issue not occur in Azure/ThingsBoard?
2. I would like us to explore/implement a non timer based approach. Cant the Cloudadapter get notified when Dispatcher agent is up and ready? Why are we even sending the cmd to Dispatcher if its not ready?

**Comment by gblewis1** [comment]
> @yengliong93 @gblewis1, I have couple of questions:
> 
> 1. I see that the changes for this change is only done for INBS cloud adapter. Could this issue not occur in Azure/ThingsBoard?

It could. Dispatcher can go up and down independently of Cloudadapter. In typical use I think dispatcher tends to be ready within a few seconds of Cloudadapter connecting to the cloud and there's a human on the other end who takes more than a few seconds to respond.

The reason Dispatcher can take a while to fully come up on boot is it has a 5 second sleep to wait for configuration agent sending it config values. This is back when we wanted flexibility to run configuration agent on a different machine. I think after that 5 second sleep we are supposed to set 'ready' in Dispatcher, but I'm not sure what the current implementation does.

> 2. I would like us to explore/implement a non timer based approach. Cant the Cloudadapter get notified when Dispatcher agent is up and ready? Why are we even sending the cmd to Dispatcher if its not ready?

I believe dispatcher is notifying cloudadapter it's ready and cloudadapter is sending the command at that point. If dispatcher's state is not set to 'ready', cloudadapter is replying to INBS with an error 'not ready'.

@yengliong93 please check my accuracy on this

**Comment by yengliong93** [comment]
@pillaiabhishek  @gblewis1  I think there are two different issues here. 

The first issue is the timing when the cloudadapter sending the request.  The fix for it is already implemented in https://github.com/intel/intel-inb-manageability/pull/527. The cloudadapter will check the status of dispatcher agent before sending the request. What Gavin mentioned above is correct. 

This is the second issue where the cloudadapter is waiting for the response from the dispatcher agent after sending the request (after confirming that the dispatcher agent is running). I took a look at the codebase and realized that the dispatcher agent only sends back the response after the operation during immediate scheduling - https://github.com/intel/intel-inb-manageability/blob/develop/inbm/dispatcher-agent/dispatcher/dispatcher_class.py#L790

If that's the case, the 10 seconds might not be enough, as the SOTA completion time might vary (possibly more than 10 seconds). 

One question here, what do we expect from the dispatcher's response if it is an immediate scheduling request?  If we only want to confirm that the immediate scheduling request has been successfully received by the dispatcher agent and the dispatcher agent has acknowledged that by sending back the acknowledgment response, the fix will be sending back the acknowledgment response after receiving the request but before running the SOTA. 

If we are expecting to receive any errors during the SOTA, like monitoring the SOTA using INBC, we should set a longer timeout in the cloudadapter to wait for the dispatcher agent to finish he SOTA.

**Comment by gblewis1** [comment]
@yengliong93 For all requests, we want to send back a response as soon as we can acknowledge. So for example, for immediate we should acknowledge first and then run the operation. For scheduled, we should schedule (because it's fast to do so), and then acknowledge. Later, in either case, we can query the job result.

**Comment by yengliong93** [comment]
@gblewis1 PR updated. The dispatcher will send back the acknowledgement response ASAP during the immediately scheduling.

File: inbm/dispatcher-agent/dispatcher/dispatcher_class.py [hunk 1]
```diff
@@ -782,12 +782,13 @@ def process_scheduled_requests(scheduled_requests: Sequence[Schedule]):
...
         for imm in schedule.immedate_requests:
+            dispatcher._send_result("", request_id)
             for manifest in imm.manifests:
                 try:
                     dispatcher.do_install(xml=manifest)
                 except (NotImplementedError, DispatcherException) as e:
-                    dispatcher._send_result(str(e), request_id)
-        dispatcher._send_result("", request_id)
+                    # TODO: Save the error for query request
+                    dispatcher._send_result(str(Result(CODE_BAD_REQUEST, str(e))))
```
Comment (gblewis1) [inline inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: _send_result needs request_id as second parameter

File: inbm/dispatcher-agent/dispatcher/dispatcher_class.py [hunk 2]
```diff
@@ -782,12 +782,13 @@ def process_scheduled_requests(scheduled_requests: Sequence[Schedule]):
             logger.debug(f"Scheduled repeated job: {repeated_schedule}")
 
         for imm in schedule.immedate_requests:
+            dispatcher._send_result("", request_id)
```
Comment (gblewis1) [inline inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: we should only send one result per request, not multiple. if all we have left are immediate requests, we should send one result with blank error and the request id first, and then process the requests.

File: inbm/dispatcher-agent/dispatcher/dispatcher_class.py (hunk 1 above)
Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: (but probably we won't have this line as we are only sending an acknowledgement; error per operation can be queried later)

Comment (yengliong93) [reply inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: I was thinking the error would be stored in the database. The user can use the query command to check the status.

Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: > I was thinking the error would be stored in the database. The user can use the query command to check the status.

Makes sense, the error should go in the Dispatcher database linked to the job.

File: inbm/dispatcher-agent/dispatcher/dispatcher_class.py (hunk 2 above)
Comment (yengliong93) [reply inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: I have moved the _send_result out of the for loop. The dispatcher will send back the acknowledgement response one time before processing the immediate scheduling.

File: inbm/dispatcher-agent/dispatcher/dispatcher_class.py (hunk 1 above)
Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/dispatcher_class.py]: Reviewing this again this morning, I believe this line `dispatcher._send_result(str(Result(CODE_BAD_REQUEST, str(e))))` should be removed as at this point we have already sent our result.

**Comment by yengliong93** [comment]
@gblewis1 PR updated. Could you have a review again?

**Review by gblewis1 — APPROVED** [review]

**Pull request merged**